src/linker/Linker/MessageOrigin.cs

```diff
@@ -39,6 +51,15 @@ public MessageOrigin (string fileName, int sourceLine = 0, int sourceColumn = 0)
 			return null;
 		}
 
+		internal static MessageOrigin? TryGetOrigin (IMemberDefinition sourceMethod, int ilOffset,
```

Comment:
I don't think this is the right place. We want the developer to be able to suppress any warning message (e.g. from 3rd party custom step). I think the whole logic should be close to MessageContainer.CreateWarning which probably will need to be expanded with additional "context" state parameter

Reply:
Absolutely agree.
The suppression should work without any special code from the place where we create the warning - with the sole exception that we require the "context" parameter. I would even go as far as make some strict rule that the context must be specified unless the code is somehow special (like internal errors, command line parsing errors and so on) - otherwise we should always be able to tell what the warning applies to (assembly or more specific).

I think from the API perspective this change should add the "context" attribute to the `CreateWarning` method (and maybe also `CreateError` for consistency). Then we should have another API which is something like `IsSuppressed` - I would probably put it onto the `MessageContainer` if it's possible, or if that doesn't work then the `LinkContext`. It should take the necessary parameters to decide if a warning will be suppressed or not. Notable it should not require the full warning message - the indented usage would be to avoid computing potentially expensive warning message if we know it's going to be suppressed.

Reply:
I've moved this logic out of `MessageOrigin` by adding the mentioned `IsSuppressed` API into the `LinkContext`, which is called from within `CreateWarningMessage`. It's now mandatory to specify a `MessageOrigin` for all created warnings, which must be either a filename or an objetct of a class implementing `IMetadataTokenProvider` (modules, types and members). The way to create warnings feels a little bit cumbersome though (`Context.LogMessage (MessageContainer.CreateWarningMessage (Context, Message, Code, MessageOrigin, [optional] Subcategory))`), I think this could be enhanced by adding new methods (`LogMessage` for info/diags, `LogError` for errors, `LogWarning` for warnings) and stop passing around the `LinkContext`.